qcfractal/postgres_harness.py

```diff
@@ -0,0 +1,287 @@
+
+        return self.is_alive(database=database_name)
+
+    def shutdown(self) -> Any:
+        """Shutsdown the current postgres instance.
+
+        """
+        ret = _run([
+            shutil.which("pg_ctl"),
+            "-D", str(self.config.database_path),
+            "stop"],
+            logger=self.logger, quiet=self.quiet) # yapf: disable
+        return ret
+
+    def initialize(self):
+        """Initializes and starts the current postgres instance.
+        """
+        if not self.quiet:
+            self.logger("Initializing the database:")
+
+        # Initialize the database
+        init_code, init_stdout = _run([shutil.which("initdb"), "-D", self.config.database_path],
+                                      logger=self.logger,
+                                      quiet=self.quiet)
+        if "Success." not in init_stdout:
+            raise ValueError(init_stdout)
+
+        # Change any configurations
+        psql_conf_file = (self.config.database_path / "postgresql.conf")
+        psql_conf = psql_conf_file.read_text()
+        if self.config.database.port != 5432:
+            assert "#port = 5432" in psql_conf
+            psql_conf = psql_conf.replace("#port = 5432", f"port = {self.config.database.port}")
+
+            psql_conf_file.write_text(psql_conf)
+
+        # Startup the server
+        start_code, start_stdout = _run([
+            shutil.which("pg_ctl"),
+            "-D", str(self.config.database_path),
+            "-l", str(self.config.database_path / self.config.database.logfile),
+            "start"],
+            logger=self.logger, quiet=self.quiet) # yapf: disable
+        if not (("server started" in start_stdout) or ("server starting" in start_stdout)):
+            raise ValueError(f"Could not start the Postgres server. Error below:\n{start_stdout}")
+
+        # Check that we are alive
+        for x in range(10):
+            if self.is_alive():
+                break
+            else:
+                time.sleep(0.1)
+        else:
+            raise ValueError(f"Could not connect to the server after booting. Boot log:\n{start_stdout}")
+
+        # Create the user and database
+        if not self.quiet:
+            self.logger(f"Building user information.")
+        _run([shutil.which("createdb"), "-p", str(self.config.database.port)])
+
+        success = self.create_database(self.config.database.default_database)
+
+        if success is False:
+            self.shutdown()
+            raise ValueError("Database created successfully, but could not connect. Shutting down postgres.")
+
+
+class TemporaryPostgres:
+    def __init__(self,
+                 database_name: Optional[str] = None,
+                 tmpdir: Optional[str] = None,
+                 quiet: bool = True,
+                 logger: 'print' = print):
+        """A PostgreSQL instance run in a temporary folder.
+
+        ! Warning ! All data is lost when the server is shutdown.
```

Comment:
I don't think this is quite true, but it may not mater. The data will be lost when the memory the `TemporaryPostgres` class points to is destroyed. Small detail, probably doesn't mater.

Reply:
Patched.